docs/architecture/adr-021-protobuf-query-encoding.md

```diff
@@ -0,0 +1,238 @@
+}
+```
+
+#### Handling of Interface Types
+
+Modules that use interface types and need true polymorphism generally force a
+`oneof` up to the app-level that provides the set of concrete implementations of
+that interface that the app supports. While app's are welcome to do the same for
+queries and implement and app-level query service, it is recommended that modules
+provide query methods that expose these interfaces via `google.protobuf.Any`.
+There is a concern on the transaction level that the overhead of `Any` is too
+high to justify its usage. However for queries this is not a concern, and
+providing generic module-level queries that use `Any` does not preclude apps
+from also providing app-level queries that return use the app-level `oneof`s.
+
+
+A hypothetical example for the `gov` module would look something like:
+
+```proto
+// x/gov/types/types.proto
+
+import "google/protobuf/any.proto";
+
+service Query {
+  rpc GetProposal(GetProposalParams) returns (AnyProposal) { }
+}
+
+message AnyProposal {
+  ProposalBase base = 1;
+  google.protobuf.Any content = 2;
+}
+```
+
+### Custom Query Implementation
+
+In order to implement the query service, we can reuse the existing [gogo protobuf](https://github.com/gogo/protobuf)
+grpc plugin, which for a service named `Query` generates an interface named
+`QueryServer` as below:
+
+```go
+type QueryServer interface {
+	QueryBalance(context.Context, *QueryBalanceParams) (*types.Coin, error)
+	QueryAllBalances(context.Context, *QueryAllBalancesParams) (*QueryAllBalancesResponse, error)
+}
+```
+
+The custom queries for our module are implemented by implementing this interface.
+
+The first parameter in this generated interface is a generic `context.Context`,
+whereas querier methods generally need an instance of `sdk.Context` to read
+from the store. Since arbitrary values can be attached to `context.Context`
+using the `WithValue` and `Value` methods, the SDK should provide a function
+`sdk.UnwrapSDKContext` to retrieve the `sdk.Context` from the provided
+`context.Context`.
+
+An example implementation of `QueryBalance` for the bank module as above would
+look something like:
+
+```go
+type Querier struct {
+	Keeper
+}
+
+func (q Querier) QueryBalance(ctx context.Context, params *types.QueryBalanceParams) (*sdk.Coin, error) {
+	balance := q.GetBalance(sdk.UnwrapSDKContext(ctx), params.Address, params.Denom)
+	return &balance, nil
+}
+```
+
+### Custom Query Registration and Routing
+
+Query server implementations as above would be registered with `AppModule`s using
+a new method `RegisterQueryServer(grpc.Server)` which could be implemented simply
+as below:
+
+```go
+// x/bank/module.go
+func (am AppModule) RegisterQueryServer(server grpc.Server) {
+	types.RegisterQueryServer(server, keeper.Querier{am.keeper})
+}
+```
+
+Underneath the hood, a new method `QueryRouter.RegisterService(sd *grpc.ServiceDesc, handler interface{})`
```

Comment:
This needs more explanation. What is a `QueryRouter` and where is it defined? What's the `handler` and it's purpose?